src/lib.rs

```diff
@@ -885,51 +925,747 @@ impl GooseAttack {
                 trace!("inner gcd: {}", u);
             }
         }
-        // 'u' will always be the greatest common divisor
-        debug!("gcd: {}", u);
+        // 'u' will always be the greatest common divisor
+        debug!("gcd: {}", u);
+
+        // Build a weighted lists of task sets (identified by index)
+        let mut weighted_task_sets = Vec::new();
+        for (index, task_set) in self.task_sets.iter().enumerate() {
+            // divide by greatest common divisor so vector is as short as possible
+            let weight = task_set.weight / u;
+            trace!(
+                "{}: {} has weight of {} (reduced with gcd to {})",
+                index,
+                task_set.name,
+                task_set.weight,
+                weight
+            );
+            let mut weighted_sets = vec![index; weight];
+            weighted_task_sets.append(&mut weighted_sets);
+        }
+
+        // Allocate a state for each user that will be hatched.
+        info!("initializing user states...");
+        let mut weighted_users = Vec::new();
+        let mut user_count = 0;
+        loop {
+            for task_sets_index in &weighted_task_sets {
+                let base_url = goose::get_base_url(
+                    self.get_configuration_host(),
+                    self.task_sets[*task_sets_index].host.clone(),
+                    self.defaults.host.clone(),
+                )?;
+                weighted_users.push(GooseUser::new(
+                    self.task_sets[*task_sets_index].task_sets_index,
+                    base_url,
+                    self.task_sets[*task_sets_index].min_wait,
+                    self.task_sets[*task_sets_index].max_wait,
+                    &self.configuration,
+                    self.metrics.hash,
+                )?);
+                user_count += 1;
+                if user_count >= self.users {
+                    trace!("created {} weighted_users", user_count);
+                    return Ok(weighted_users);
+                }
+            }
+        }
+    }
+
+    // Configure which mode this GooseAttack will run in.
+    fn set_attack_mode(&mut self) -> Result<(), GooseError> {
```

Comment:
@LionsAd I like having an `attack_mode`, but this could just live in the configuration struct if that's somehow better.